### PR TITLE
model forwards can take an inputs_embeds param

### PR DESCRIPTION
**Pull request by julien-c**

### CODE REVIEW

File: transformers/modeling_gpt2.py [hunk 1]
```diff
@@ -368,9 +368,15 @@ def _prune_heads(self, heads_to_prune):
...
-        input_shape = input_ids.size()
-        input_ids = input_ids.view(-1, input_shape[-1])
+    def forward(self, input_ids=None, past=None, attention_mask=None, token_type_ids=None, position_ids=None, head_mask=None, inputs_embeds=None):
+        if input_ids is not None:
+            input_shape = input_ids.size()
+            input_ids = input_ids.view(-1, input_shape[-1])
+        elif inputs_embeds is not None:
+            input_shape = inputs_embeds.size()[:-1]
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
```
Comment (thomwolf) [inline transformers/modeling_gpt2.py]: Should we raise an error if both are provided?
If not we should be explicit about which one has priority in the docstring (this input should be in the docstring anyway).

File: transformers/modeling_gpt2.py [hunk 2]
```diff
@@ -382,8 +388,9 @@ def forward(self, input_ids, past=None, attention_mask=None, token_type_ids=None
         else:
             past_length = past[0][0].size(-2)
         if position_ids is None:
-            position_ids = torch.arange(past_length, input_ids.size(-1) + past_length, dtype=torch.long, device=input_ids.device)
-            position_ids = position_ids.unsqueeze(0).expand_as(input_ids)
+            device = input_ids.device if input_ids is not None else inputs_embeds.device
+            position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
```
Comment (thomwolf) [inline transformers/modeling_gpt2.py]: 👍

**Review by thomwolf — APPROVED** [review]
Ok, for me, just check the comment and add the input in the docstring.
Also, I think we should add this input in all the models.

File: transformers/modeling_gpt2.py (hunk 1 above)
Comment (julien-c) [reply transformers/modeling_gpt2.py]: indeed

File: transformers/modeling_transfo_xl.py [hunk 3]
```diff
@@ -693,15 +697,22 @@ def _update_mems(self, hids, mems, qlen, mlen):
...
         # the original code for Transformer-XL used shapes [len, bsz] but we want a unified interface in the library
         # so we transpose here from shape [bsz, len] to shape [len, bsz]
-        input_ids = input_ids.transpose(0, 1).contiguous()
+        if input_ids is not None and inputs_embeds is not None:
+            raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
+        elif input_ids is not None:
+            input_ids = input_ids.transpose(0, 1).contiguous()
+            qlen, bsz = input_ids.size()
+        elif inputs_embeds is not None:
+            inputs_embeds = inputs_embeds.transpose(0, 1).contiguous()
```
Comment (thomwolf) [inline transformers/modeling_transfo_xl.py]: 👍

File: transformers/tests/modeling_common_test.py [hunk 4]
```diff
@@ -525,6 +525,19 @@ def check_same_values(layer_1, layer_2):
...
+            config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+            input_ids = inputs_dict["input_ids"]
+            del inputs_dict["input_ids"]
+
+            for model_class in self.all_model_classes:
+                model = model_class(config)
+                model.eval()
+
+                wte = model.get_input_embeddings()
+                inputs_dict["inputs_embeds"] = wte(input_ids)
```
Comment (thomwolf) [inline transformers/tests/modeling_common_test.py]: perfect

**Comment by thomwolf** [comment]
LGTM.
Feel free to add the TF version or merge if you don't want to add them now.

**Pull request merged**